File: 0271-encode-and-decode-strings/0271-encode-and-decode-strings.js
/**
 * Encodes a list of strings to a single string.
 *
 * @param {string[]} strs
 * @return {string}
 */

// Input: dummy_input = ["Hello#","World"]
// "#5Hello#5World"
// Hello + #(indicate end/beg of a string) + 5(length of a string)
// 
// time n
// space of n res return res
//
//
var encode = function(strs) {//["Hello","World"]
    let res = ""
    for (let s of strs) {
        len = s.length //10
        res += len
        res += "#"
        res += s
    }
    return res
};

/**
 * Decodes a single string to a list of strings.
 *
 * @param {string} s
 * @return {string[]}
 */
var decode = function(s) {//5#Hello"5#World"
    //0#2#vn
    let res = [] 
    let i = 0;
    
    while(i < s.length){
        let j = i;
        while(s[j] !== "#"){
            j++;
        }
        let length = parseInt(s.substring(i,j))
        i = j + 1
        j = i + length
        string = s.substring(i, j)
        res.push(string)
        i = j
    }
    
    return res
};

/**
 * Your functions will be called as such:
 * decode(encode(strs));
 */

File: 0271-encode-and-decode-strings/0271-encode-and-decode-strings.py
#

class Codec:
    def encode(self, strs: List[str]) -> str:
        """Encodes a list of strings to a single string.
        """
        res = ""
        for s in strs:#hello, world
            res += str(len(s)) + "#" + s # 5#Hello
        
        return res

        

    def decode(self, s: str) -> List[str]:
        """Decodes a single string to a list of strings
        """
        res = []
        i = 0
        while i < len(s):
            j = i
            while s[j] != "#":
                j += 1
            length = int(s[i:j])#get lenght in form of int(4 or 10)
            res.append(s[j + 1: j + 1 +length]) #slice the j+1 start to j+1+length end
            i = j + 1 + length #move i pointer

        return res


# Your Codec object will be instantiated and called as such:
# codec = Codec()
# codec.decode(codec.encode(strs))